floyd/cli/auth.py

```diff
@@ -9,22 +11,125 @@
 from floyd.model.credentials import Credentials
 from floyd.log import logger as floyd_logger
 
+import threading
+import webbrowser
+import subprocess
+if sys.version_info[0] < 3:
+    import urlparse
+    from BaseHTTPServer import (
+        BaseHTTPRequestHandler,
+        HTTPServer,
+    )
+    from Queue import Queue
+else:
+    from urllib import parse as urlparse
+    from http.server import (
+        BaseHTTPRequestHandler,
+        HTTPServer,
+    )
+    from queue import Queue
+
+
+class LoginServer(HTTPServer, object):
+    def __init__(self, server_address, RequestHandlerClass, key_queue):
+        super(LoginServer, self).__init__(server_address, RequestHandlerClass)
+        self.key_queue = key_queue
+
+
+class LoginHttpRequestHandler(BaseHTTPRequestHandler):
+    def do_OPTIONS(self):
+        self.send_response(200, "ok")
+        self.send_header('Access-Control-Allow-Origin', '*')
+        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
+        self.end_headers()
+
+    def do_GET(self):
+        params = urlparse.parse_qs(urlparse.urlparse(self.path).query)
+        key = params.get('apikey')
+        if not key:
+            self.send_response(400)
+            return
+
+        self.server.key_queue.put(key[0])
+
+        self.send_response(200)
+        self.send_header('Access-Control-Allow-Origin', '*')
+        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
+        self.send_header('Content-type', 'text/html')
+        self.end_headers()
+        self.wfile.write("OK".encode('utf-8'))
+
+    def log_message(self, fmt, *args):
+        return
+
+
+@contextlib.contextmanager
+def get_free_port():
+    try:
+        s = socket.socket()
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        s.bind(('localhost', 0))
+        yield s.getsockname()[1]
+        s.close()
+    except socket.error:
+        yield None
+
+
+def wait_for_apikey():
+    key_queue = Queue()
+    with get_free_port() as port:
+        if not port:
+            return
+        server = LoginServer(('', port), LoginHttpRequestHandler, key_queue)
+
+    t = threading.Thread(
+        target=server.serve_forever)
+    t.daemon = True
+    t.start()
+
+    cli_host = 'http://127.0.0.1'
+    url = '%s/cli_login?callback=%s:%s' % (floyd.floyd_web_host, cli_host, port)
+    subprocess.check_output(
+        [sys.executable, '-m', 'webbrowser', url], stderr=subprocess.STDOUT)
+
+    floyd_logger.info('waiting for login from browser...')
+    while True:
+        try:
+            apikey = key_queue.get(timeout=500)
```

Comment:
What is the unit of the timeout here, seconds?

Since this is a while loop, does this mean the cli will continue to wait even if the browser key handoff fails for some reason?